Replaced valuecheckMatrix() with CompareMatrices(). Updated some unit tests use Google Test.

liangfok opened the pull request:
Closes https://github.com/RobotLocomotion/drake/issues/1916.

Getting this merged is a prerequisite for https://github.com/RobotLocomotion/drake/pull/1898.

liangfok commented:
This was reviewed by @david-german-tri in ReviewBoard: https://rbcommons.com/s/tri/r/2/

david-german-tri commented:
There are many more files in this PR!  I notice that you're updating many of the tests to use the CompareMatrices function instead of the EigenMatrixIsApproximatelyEqualMatcher?

liangfok commented:
> There are many more files in this PR!

Hmm, did not all of the files show up in ReviewBoard?

liangfok commented:
> I notice that you're updating many of the tests to use the CompareMatrices function instead of the EigenMatrixIsApproximatelyEqualMatcher?

I figured that semantically, the following two statements are identical for matrices `x` and `y`:

1. `EXPECT_THAT(x, drake::test::EigenMatrixIsApproximatelyEqual( y, tolerance, MatrixCompareType::absolute));`
2. `EXPECT_TRUE(CompareMatrices(x, y, tolerance, MatrixCompareType::absolute));`

I went with the second option since it was more of a direct drop-in replacement for the old `valuecheckMatrix()`.

If the Google Mock version is better, maybe we can switch to the Google Mock version in a future issue?

liangfok commented:
test this please

liangfok commented:
The Jekins CI server is [reporting some errors](https://drake-jenkins.csail.mit.edu/job/experimental/447/) with this PR. They are summarized below.

**msvc-32:**
```
vector_test.obj : error LNK2019: unresolved external symbol "class testing::internal::Mutex testing::internal::g_linked_ptr_mutex" (?g_linked_ptr_mutex@internal@testing@@3VMutex@12@A) referenced in function "private: void __cdecl testing::internal::linked_ptr<class testing::MatcherInterface<class Eigen::Matrix<double,2,1,0,2,1> const &> const >::depart(void)" (?depart@?$linked_ptr@$$CBV?$MatcherInterface@AEBV?$Matrix@N$01$00$0A@$01$00@Eigen@@@testing@@@internal@testing@@AEAAXXZ) [C:\projects\workspace\experimental\compiler\msvc-64\label\windows\drake\pod-build\core\test\vector_test.vcxproj]
```

**msvc-64:**
```
vector_test.obj : error LNK2019: unresolved external symbol "class testing::internal::Mutex testing::internal::g_linked_ptr_mutex" (?g_linked_ptr_mutex@internal@testing@@3VMutex@12@A) referenced in function "private: void __cdecl testing::internal::linked_ptr<class testing::MatcherInterface<class Eigen::Matrix<double,2,1,0,2,1> const &> const >::depart(void)" (?depart@?$linked_ptr@$$CBV?$MatcherInterface@AEBV?$Matrix@N$01$00$0A@$01$00@Eigen@@@testing@@@internal@testing@@AEAAXXZ) [C:\projects\workspace\experimental\compiler\msvc-64\label\windows\drake\pod-build\core\test\vector_test.vcxproj]
```

**clang,linux:**
```
[xUnit] [ERROR] - Test reports were found but not all of them are new. Did all the tests run?
```

I'm not sure how to debug the msvc-32 and mdvc-64 problems without a Windows machine, but the clang error appears to be a problem with Jenkins that was already resolved judging by the comment [here](https://github.com/RobotLocomotion/drake/pull/1945#issuecomment-202507889).

liangfok commented:
After merging with the latest from RobotLocomotion/drake's head and re-testing, the `clang,linux` test now passes, but the two Windows tests still fail:

*msvc-32:*

```
C:\projects\workspace\experimental\compiler\msvc-32\label\windows\build\include\gtest/internal/gtest-internal.h(892): error C2718: 'const T1': actual parameter with requested alignment of 16 won't be aligned [C:\projects\workspace\experimental\compiler\msvc-32\label\windows\drake\pod-build\core\test\vector_test.vcxproj]
```

*msvc-64:*

```
vector_test.obj : error LNK2019: unresolved external symbol "class testing::internal::Mutex testing::internal::g_linked_ptr_mutex" (?g_linked_ptr_mutex@internal@testing@@3VMutex@12@A) referenced in function "private: void __cdecl testing::internal::linked_ptr<class testing::MatcherInterface<class Eigen::Matrix<double,2,1,0,2,1> const &> const >::depart(void)" (?depart@?$linked_ptr@$$CBV?$MatcherInterface@AEBV?$Matrix@N$01$00$0A@$01$00@Eigen@@@testing@@@internal@testing@@AEAAXXZ) [C:\projects\workspace\experimental\compiler\msvc-64\label\windows\drake\pod-build\core\test\vector_test.vcxproj]
```

@david-german-tri, these seem like issues linking Google Test libraries on Windows. Any ideas?

david-german-tri commented on drake/examples/Acrobot/test/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 if (LCM_FOUND)
```
I doubt this explains your MSVC problem, but you probably want `AND GTEST_FOUND` here

liangfok replied:
Thanks I added the missing `GTEST_FOUND` guards.

jamiesnape commented:
retest this please

liangfok commented:
CI tests on `drake005.csail.mit.edu` and `drake006.csail.mit.edu` are failing but no additional information is provided:

![screen shot 2016-03-28 at 11 15 55 pm](https://cloud.githubusercontent.com/assets/1388098/14097030/56844054-f53b-11e5-93c6-a6bd971d64cd.png)

![screen shot 2016-03-28 at 11 16 44 pm](https://cloud.githubusercontent.com/assets/1388098/14097033/5ac2a516-f53b-11e5-9e8b-8df356f8645f.png)

Not sure how to proceed to debug.

billhoffman commented:
It is also failing on Windows: https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=3676.  The drake005 and drake006 issues look like system problems with testing results @RussTedrake.

liangfok commented:
Thanks @billhoffman. I don't have a Windows machine yet so I can't replicate the error. Thus, I'm first going to see if simply disabling alignment using Eigen's `EIGEN_DONT_ALIGN` macro makes this compiler error go away: https://github.com/RobotLocomotion/drake/pull/1953/commits/004787cfccb482b62fd7433b3c32ff9b6dfaf409.

liangfok commented:
Unfortunately, using Eigen's `EIGEN_DONT_ALIGN` did not help. The [error on Windows ](https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=3792)is the same as before.

Possible relevant info [here](http://stackoverflow.com/questions/1281415/error-c2719-val-formal-parameter-with-declspecalign16-wont-be-alig).

sherm1 commented:
> Unfortunately, using Eigen's EIGEN_DONT_ALIGN did not help.

I think that error means that alignment was still being requested. Maybe the macro wasn't in the right place to take effect?

liangfok commented:
@RussTedrake looks like drake005 is still down. Or, is my code resulting in such bad compilation errors that drake005 isn't even able to report them?

liangfok commented:
@sherm1: Thanks for the hint. I eventually isolated the problem to be within my Google Mock component and worked around it by replacing it with a simpler Google Test check. This is possible since both methods end up calling the same method. Here is the relevant commit: https://github.com/RobotLocomotion/drake/pull/1953/commits/0d32327a962e483b2df0af648270567592a44ae1

sherm1 commented:
> I eventually isolated the problem to be within my Google Mock component and worked around it by replacing it with a simpler Google Test check

Great -- I like that `CompareMatrices` function, especially the relative tolerance code.

liangfok commented:
@david-german-tri I think this one is ready to be merged despite the errors on drake005 and drake006. Can you review it?

david-german-tri commented on .gitignore:
```diff
@@ -58,3 +58,4 @@ drake/doc/DoxygenCPP
 *.bak
 Pendulum.pmd
 drake/doc/_build
+.reviewboardrc
```
Let's not do this, since the Review Board experiment is pretty well dead.

liangfok replied:
OK.

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+#include "drake/util/eigen_matrix_compare.h"
```
This file  should be named eigen_matrix_compare_test.h

https://google.github.io/styleguide/cppguide.html#File_Names

liangfok replied:
OK.

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+using drake::util::MatrixCompareType;
+
+namespace drake {
+namespace test {
```
Why not `namespace util` instead, and drop the `using`-decl above?

liangfok replied:
I didn't want to pollute the `drake::util` namespace and thought "test" better describes the purpose of this class compared to an anonymous namespace. Should I use three levels of namespaces instead?
```
namespace drake {
namespace util {
namespace test {
```

liangfok replied:
Actually, I'm now leaning towards the following to be consistent with the directory structure.

```
namespace drake {
namespace core {
namespace test {
```

liangfok replied:
Nevermind. Since `eigen_matrix_compare_test.cc` specifically evaluates a function defined in `drake/util/eigen_matrix_compare.h`, I will move `eigen_matrix_compare_test.cc` into `drake/util/test/` and use:

```
namespace drake {
namespace util {
namespace test 
```

david-german-tri replied:
Yes.  I still think an anonymous namespace is preferable to `namespace test`, but apparently there's no section of the style guide I can cite for that view, so I can live with your solution.

liangfok replied:
@david-german-tri: I decided to take your suggestion and use an anonymous namespace to further reduce the number of symbols exported by Drake:

```
namespace drake {
namespace util {
namespace {
```

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+  Eigen::MatrixXd m3(2, 2);
+  m3 << 100, 200, 300, 400;
+
+  double tolerance = 1e-8;
```
const (here and below)

liangfok replied:
OK.

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+  // They should be considered different.
+  EXPECT_FALSE(CompareMatrices(m1, m4, tolerance, MatrixCompareType::absolute,
+                               &error_msg))
+      << "m1 =\n"
```
Did you mean to delete these lines?

liangfok replied:
Yes, thanks.

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+  EXPECT_FALSE(CompareMatrices(m1, m2, one_pct, MatrixCompareType::relative,
+                               &error_msg));
+
+  // The difference between m1 and m2 is equal to 1%.
```
10%

liangfok replied:
Fixed. Thanks.

david-german-tri commented on drake/core/test/matrix_compare_test.cc:
```diff
@@ -0,0 +1,152 @@
+  Eigen::MatrixXd m2(2, 2);
+  m2 << 100, 100 * 0.9, 100, 100;
+
+  double one_pct = 0.01;
```
Naming these variables doesn't seem like much of a win to me.  (optional)

liangfok replied:
I removed the variables. I usually go by the rule that the fewer magic numbers present the better. However, in this case, since the variable is only used once as a method's parameter and the name of the parameter is "tolerance", I'll make an exception to this magic number rule.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+#include "drake/examples/Pendulum/Pendulum.h"  // to get some types
```
nit: alphabetize the #includes

liangfok replied:
Fixed.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+  state.theta = 0.2;
+  state.thetadot = .3;
+
+  EXPECT_EQ(size(state), static_cast<size_t>(2));
```
is the static_cast necessary?

liangfok replied:
I removed the static cast and this unit test continued to work on both Ubuntu Linux and OS X. I haven't finished spinning up my Windows development platform but will assume the static cast is not necessary and rely on the CI servers to detect and invalidate this assumption.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+  EXPECT_EQ(size(state), static_cast<size_t>(2));
+
+  state = x;
+  EXPECT_EQ(state.thetadot, 0.4);
```
Why not check theta also?

liangfok replied:
Added a check for theta.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+
+// Tests the RowsAtCompileTime
+TEST(VectorTest, RowsAtCompileTime) {
+  EXPECT_TRUE((Eigen::Matrix<double, 2, 1>::RowsAtCompileTime == 2))
```
EXPECT_EQ(2, ...)

liangfok replied:
Fixed.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+      << "zero is arbitrary";
+}
+
+// Verify that the least common ancestor of the I/O relations
```
super nit: Verifies (here and below)

liangfok replied:
Fixed.

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,143 @@
+  InputOutputRelation g(InputOutputRelation::Form::LINEAR);
+  InputOutputRelation f(InputOutputRelation::Form::POLYNOMIAL);
+
+  EXPECT_TRUE(InputOutputRelation::composeWith(g, f).form ==
```
EXPECT_EQ

at which point the "should be poly" message will be redundant, here and below

liangfok replied:
Fixed.

david-german-tri commented on drake/solvers/test/testMobyLCP.cpp:
```diff
@@ -174,10 +186,14 @@ TEST(testMobyLCP, testProblem3) {
   // Local and Global Optimization".
   Eigen::Matrix<double, 3, 3> M;
+
+  // clang-format off
```
Useful tip, I didn't know about this.  Thanks!

david-german-tri commented on drake/solvers/test/testOptimizationProblem.cpp:
```diff
@@ -3,9 +3,34 @@
 #include "drake/solvers/Optimization.h"
```
alphabetize

liangfok replied:
Done.

david-german-tri commented on drake/solvers/test/testOptimizationProblem.cpp:
```diff
@@ -65,28 +90,44 @@ void trivialLeastSquares() {
-  valuecheckMatrix(b, x.value(), 1e-10);
+  EXPECT_TRUE(
+      CompareMatrices(b, x.value(), 1e-10, MatrixCompareType::absolute));
+  // valuecheckMatrix(b, x.value(), 1e-10);
```
remove these

liangfok replied:
Done.

david-german-tri commented on drake/systems/plants/test/testAccelerometer.cpp:
```diff
@@ -1,78 +1,105 @@
+using Drake::RigidBodyAccelerometer;
+
+namespace drake {
+namespace test {
```
Oh, I see (belatedly): you're using the `test` namespace to protect the production namespaces from pollution.  Great idea, but why not just use an anonymous namespace?

liangfok replied:
Agreed anonymous namespace is better since that is completely hidden from external users of the library. Switched to use it.

david-german-tri commented on drake/systems/plants/test/testGyroscope.cpp:
```diff
@@ -1,37 +1,56 @@
 #include "drake/systems/plants/RigidBodySystem.h"
```
alphabetize

liangfok replied:
Fixed.

david-german-tri commented on drake/systems/plants/test/testGyroscope.cpp:
```diff
@@ -1,37 +1,56 @@
+    EXPECT_TRUE(CompareMatrices(getGyroscopeOutput(rigid_body_sys, ang_vel),
+                                ang_vel, tol, MatrixCompareType::absolute));
+
+    // valuecheckMatrix(getGyroscopeOutput(rigid_body_sys, ang_vel),
```
delete

liangfok replied:
Done.

david-german-tri commented on drake/util/test/testPolynomial.cpp:
```diff
@@ -1,35 +1,53 @@
   Polynomial<CoefficientType> poly(coefficients);
 
-  cout << poly << endl;
+  // cout << poly << endl;
```
remove

liangfok replied:
Done.

david-german-tri commented on drake/util/test/testPolynomial.cpp:
```diff
@@ -153,11 +171,17 @@ void testPolynomialMatrix() {
 }
 
-int main(int argc, char **argv) {
+TEST(PolynomialTest, IntegrateAndDerivative) {
```
Integral

liangfok replied:
Fixed.

david-german-tri commented on drake/util/testUtil.h:
```diff
@@ -42,77 +42,6 @@ void valuecheck(const T& a, const T& b, std::string error_msg = "") {
- * }
- */
-template <typename DerivedA, typename DerivedB>
-void valuecheckMatrix(const Eigen::MatrixBase<DerivedA>& a,
```
Hooray!!

david-german-tri commented:
Added a bunch of minor comments.  Obviously, many of the tests you've migrated are still in a clunky state, but I didn't pile comments onto those - this change is a huge improvement!

liangfok commented:
@david-german-tri: Thanks for the thorough code review. I believe I addressed all of your comments. Can you verify the changes and merge if everything is acceptable?

david-german-tri commented on drake/core/test/vector_test.cc:
```diff
@@ -0,0 +1,142 @@
+              InputOutputRelation::Form::POLYNOMIAL);
+
+  EXPECT_EQ(InputOutputRelation::composeWith(f, g).form,
+              InputOutputRelation::Form::POLYNOMIAL);
```
Looks like everything should be clang-formatted again

liangfok replied:
@david-german-tri: I just ran clang-format and noticed it may be violating the style guide's specification on the order of includes. Here is an example change made by clang-format:

![screen shot 2016-03-31 at 3 00 24 pm](https://cloud.githubusercontent.com/assets/1388098/14187584/c9ff1c8a-f751-11e5-9d50-1bf48ab2b816.png)

Shouldn't `gtest.h` be considered "other libraries' .h" meaning it should go before drake's header files?

jwnimmer-tri replied:
Is the problem that you should use `<gtest/gtest.h>` and not `"gtest/gtest.h"`?

david-german-tri commented on drake/examples/Acrobot/test/urdfDynamicsTest.cpp:
```diff
@@ -1,19 +1,34 @@
+
+namespace drake {
+namespace examples {
+namespace Acobot {
```
acrobot

liangfok commented:
The CI servers found a problem that I failed to detect due to the fact that I did not have Matlab installed on any of my local machines:

```
/home/ubuntu/workspace/experimental/compiler/gcc/label/linux_matlab/drake/systems/plants/test/compareParsersmex.cpp: In function ‘void mexFunction(int, mxArray**, int, const mxArray**)’:
/home/ubuntu/workspace/experimental/compiler/gcc/label/linux_matlab/drake/systems/plants/test/compareParsersmex.cpp:104:64: error: ‘valuecheckMatrix’ was not declared in this scope
       valuecheckMatrix(matlab_H, cpp_H, 1e-8, "H doesn't match");
                                                                ^
/home/ubuntu/workspace/experimental/compiler/gcc/label/linux_matlab/drake/systems/plants/test/compareParsersmex.cpp:116:55: error: ‘valuecheckMatrix’ was not declared in this scope
                        "joint_limit_min doesn't match");
                                                       ^
/home/ubuntu/workspace/experimental/compiler/gcc/label/linux_matlab/drake/systems/plants/test/compareParsersmex.cpp:125:70: error: ‘valuecheckMatrix’ was not declared in this scope
       valuecheckMatrix(matlab_phi, cpp_phi, 1e-8, "phi doesn't match");
                                                                      ^
```

jwnimmer-tri commented:
Here's a final suggestion, taken from the changelog -- for any users who consult this PR for advice:

Unit tests that originally contained
```
    valuecheckMatrix(matrix1, matrix2, tolerance);
```
should be modified to use Google Test and the following line:
```
    EXPECT_TRUE(drake::util::CompareMatrices(matrix1, matrix2, tolerance, MatrixCompareType::absolute));
```
